proposals/0000-all-threads-syscall.rst

```diff
@@ -0,0 +1,146 @@
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/533>`_.
+.. sectnum::
+.. contents::
+
+Several Linux kernel security features exposed as syscalls are applied to the
+thread invoking such syscall only, then inherited by newly created threads.
+Since the GHC threaded RTS creates multiple threads before Haskell code is
+executed, we need a way to invoke some syscall in all running threads.
+
+
+Motivation
+----------
+In Linux, many attributes of the context in which code runs is kept per thread
+instead of process-global, as often assumed. As an example, the ``setuid``
+syscall will only change the UID of the thread invoking the syscall, not of
+all threads running in the process, as required by POSIX.
+
+This is often not an issue because the ``libc`` implementation of such library
+functions jump through hoops to ensure the syscalls are executed in every
+running thread before returning, effectively guaranteeing POSIX compliance.
+This is known as the ``setxid`` functionality, which is internal to the libc
+library.
+
+As such, when calling the (C library) ``setuid`` function from within Haskell
+code will result in the desired result: all code in the process is now running
+using the new UID.
+
+Several Linux syscalls don't have ``libc`` wrappers (yet), or aren't applied
+using ``setxid``-like mechanisms. In C codebases, a developer would ensure
+all attributes are set by code running in the main thread before any new
+threads are spawned. However, in software using some managed runtime system,
+said runtime may spawn threads before application code is launched, which is
+the case when using the GHC threaded RTS.
+
+There are work-arounds available (see below), though these are brittle and rely
+on RTS implementation details. Hence, other platforms (e.g., Go) introduced
+the ability to invoke a syscall in all (managed) threads.
+
+Proposed Change Specification
+-----------------------------
+On Linux, the GHC threaded RTS will be extended with a new function::
+
+    long rts_allThreadsSyscall(long num,
+                               long arg1, long arg2, long arg3,
+                               long arg4, long arg5, long arg6)
+
+This function will promptly invoke the given syscall ``num`` with the given
+args in all (POSIX) threads managed by the RTS.
+
+The semantics of this function are the same as those of |glibc|_
+|nptl_setxid|_ or Go_'s |syscall_runtime_doAllThreadsSyscall|_:
+
+- First, the syscall is invoked in the calling thread. Keep track of the
+  result.
+
+- Then, set up machinery to invoke the syscall in all (other) running threads,
+  and wait for them to accomplish this task, blocking the calling thread.
+
+- If during the invocation of the syscall in any of the other threads the result
+  is *not* equal to the first invocation, abort the process.
+
+- Return the result.
```

Comment:
Couldn't we have a more generic interface that doesn't mention syscalls at all? E.g.

```C
// callback for runOnAllThreads
// return false to abort before all threads have run the callback function
typedef bool (*run_callback)(void *user_data);

// run the given callback on all threads sequentially until one of them returns false
// or all of them have executed the callback.
void rts_runOnAllThreads(run_callback cb, void * user_data);
```

The business of keeping the result etc. would be done in the callback using the given `user_data` as storage.
By doing this, platform specific syscall calling could be done in a library instead of in the RTS.

Reply:
Thanks for the feedback! Let me break this up in two parts:

- Running the action (whichever it may be, only syscalls or generic callbacks) in every thread sequentially should work, however, I'm not sure why that'd be desirable (vs. running them concurrently): once it turns out one of them returns "something else" (or `false`), the only way forward is `abort()` since the process would be in a very weird state. Hence, running them sequentially would only cause latency without other benefits?
- On allowing a generic callback to be passed in, this may seem enticing, indeed. However, with great power comes great responsibility. Depending on the implementation of the feature inside the RTS, one is very limited as to what one is allowed to do within such callback. In the design of the Glibc or Go runtime implementation of a similar feature, a signal and signal handler is used to interrupt running threads to perform the syscall. As such, the given callback would run inside a signal handler, and hence only async-signal-safe functions may be used (in the callback), a limitation the RTS itself can't enforce.

To conclude: I wouldn't oppose against using a more generic approach, if we deem the (non-enforcable) limitations of what can be done in the callback acceptable (I fail to see any other use-cases but triggering a syscall, though). As for running sequentially across threads, I'm not sure what the benefit would be.

Reply:
Indeed we could run the callback in parallel. We should document this properly as users would then have to use atomic/locks to deal with their `user_data`. Not sure why I proposed to do it sequentially.

I don't think we would use a signal to implement this. We already have a thread synchronization mechanism (e.g. for major GC) so we could reuse it. We would have to handke threads not used for running capabilities, such as the ticker. Not sure about FFI threads: we could interrupt the interruptible ones and fail if there are non-interruptible ones running. In any case, we would have to document what's possible to do in the callback (blocking, etc.).
 
The generic approach seems better to me but you could mention both in the proposal as alternative implementations. Note that the generic approach has an interface similar to `rts_listThreads` which has been added in 9.6.

Reply:
> By doing this, platform specific syscall calling could be done in a library instead of in the RTS.

This alone is a huge benefit, but baking in less OS-specific stuff when an abstract interface will do is also a bonus to me.

Reply:
Sorry folks, this didn't drop off my radar. I'll integrate the comments shortly.